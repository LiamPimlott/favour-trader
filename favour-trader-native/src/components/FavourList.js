import React, { Component } from 'react';
import { FlatList, StyleSheet, Text, View } from 'react-native';
import { Button, List } from 'react-native-elements';
import FavourListItem from './FavourListItem.js';

export default class FavourList extends Component {
    constructor() {
        super();
        this.state = {
            selected: (new Map()),
        };
    }

    componentDidMount() {
        if (this.props.initialSelected.length !== 0) {
            this.setState({
                selected: this.props.initialSelected,
            });
        }
    }

    _keyExtractor = (item, index) => item._id;

    _onPressItem = (id) => {
        // updater functions are preferred for transactional updates
        this.setState((state) => {
            // copy the map rather than modifying state.
            const selected = new Map(state.selected);
            selected.set(id, !selected.get(id)); // toggle
            return { selected };
        });
    };

    _onSubmit = () => {
        this.props.onSubmit(this.state.selected);
    }

    _onCancel = () => {
        this.props.onCancel();
    }

    _renderItem = ({ item }) => (
        <FavourListItem
            id={item}
            onPressItem={this._onPressItem}
            selected={!!this.state.selected.get(item)}
            title={item.category.skill}
            subtitle={item.description}
        />
    );

    render() {
        return (
            <View>
                <Text>{this.props.title}</Text>
                <List style={styles.container}>
                    <FlatList
                        data={this.props.data}
                        extraData={this.state}
                        keyExtractor={this._keyExtractor}
                        renderItem={this._renderItem}
                    />
                </List>
                <View style={styles.buttonContainer}>
                    <Button
                        title={'← Previous'}
                        onPress={this._onCancel}
                        buttonStyle={{
                            backgroundColor: "rgba(92, 99,216, 1)",
                            width: 125,
                            height: 45,
                            borderColor: "transparent",
                            borderWidth: 0,
                            borderRadius: 5
                        }}
                        containerStyle={{ marginTop: 20 }}
                    />
                    <Button
                        title={'Continue →'}
                        onPress={this._onSubmit}
                        buttonStyle={{
                            backgroundColor: "rgba(92, 99,216, 1)",
                            width: 125,
                            height: 45,
                            borderColor: "transparent",
                            borderWidth: 0,
                            borderRadius: 5
                        }}
                        containerStyle={{ marginTop: 20 }}
                    />
                </View>
            </View>
        );
    }
}

const styles = StyleSheet.create({
    container: {
        flex: 1,
        backgroundColor: '#7851a9',
        alignItems: 'center',
        justifyContent: 'center',
    },
    buttonContainer: {
        paddingLeft: 10,
        paddingRight: 10,
        flexDirection: 'row',
        justifyContent: 'space-between',
        marginTop: 10,
    },
});