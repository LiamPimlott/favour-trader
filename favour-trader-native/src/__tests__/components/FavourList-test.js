import 'react-native';
import React from 'react';
import { shallow, configure } from 'enzyme';
import moxios from "moxios";
import FavourList from '../../components/FavourList.js';
import toJson from 'enzyme-to-json'; //added this line

beforeEach(function () {
    const Adapter = require("enzyme-adapter-react-16");
    configure({adapter: new Adapter()});
    moxios.install();
});

describe('Testing FavourList component', () => {
    it('renders as expected', () => {
        const length = 5;
        const title =`What will you do for Test user?`;
        const data ='Test skill 1';
        const onCancel=() => navigation;
        const onSubmit= {onSubmit:jest.fn()};
        const initialSelected='Test skill 2';

        const wrapper = shallow(
            <FavourList length={length} title={title} data={data} onCancel={onCancel}
                        onSubmit={onSubmit} initialSelected={initialSelected}/>
        );
        expect(toJson(wrapper)).toMatchSnapshot(); //edited this line
    });
});