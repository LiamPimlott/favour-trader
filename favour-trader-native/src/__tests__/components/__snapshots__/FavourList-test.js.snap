// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing FavourList component renders as expected 1`] = `
<View>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
  >
    What will you do for Test user?
  </Text>
  <List
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#7851a9",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <FlatList
      data="Test skill 1"
      disableVirtualization={false}
      extraData={
        Object {
          "selected": "Test skill 2",
        }
      }
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      numColumns={1}
      onEndReachedThreshold={2}
      renderItem={[Function]}
      scrollEventThrottle={50}
      updateCellsBatchingPeriod={50}
      windowSize={21}
    />
  </List>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginTop": 10,
        "paddingLeft": 10,
        "paddingRight": 10,
      }
    }
  >
    <Button
      TouchableComponent={[Function]}
      buttonStyle={
        Object {
          "backgroundColor": "rgba(92, 99,216, 1)",
          "borderColor": "transparent",
          "borderRadius": 5,
          "borderWidth": 0,
          "height": 45,
          "width": 125,
        }
      }
      clear={false}
      containerStyle={
        Object {
          "marginTop": 20,
        }
      }
      disabled={false}
      iconRight={false}
      loadingProps={
        Object {
          "color": "white",
          "size": "small",
        }
      }
      onPress={[Function]}
      title="← Previous"
    />
    <Button
      TouchableComponent={[Function]}
      buttonStyle={
        Object {
          "backgroundColor": "rgba(92, 99,216, 1)",
          "borderColor": "transparent",
          "borderRadius": 5,
          "borderWidth": 0,
          "height": 45,
          "width": 125,
        }
      }
      clear={false}
      containerStyle={
        Object {
          "marginTop": 20,
        }
      }
      disabled={false}
      iconRight={false}
      loadingProps={
        Object {
          "color": "white",
          "size": "small",
        }
      }
      onPress={[Function]}
      title="Continue →"
    />
  </View>
</View>
`;
